File: BAEKJOON/1450 - 냅색문제/냅색문제.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program
{
    // 부분 집합의 합을 계산하는 함수
    static List<long> GetSubsets(List<int> weights)
    {
        int n = weights.Count;
        List<long> subsets = new List<long>();

        // 모든 부분 집합을 계산하여 그 합을 저장
        for (int i = 0; i < (1 << n); i++)
        {
            long sum = 0;
            for (int j = 0; j < n; j++)
            {
                if ((i & (1 << j)) != 0)
                {
                    sum += weights[j];
                }
            }
            subsets.Add(sum);
        }

        return subsets;
    }

    // 가능한 조합의 수를 계산하는 함수
    static long CountValidCombinations(int N, long C, List<int> weights)
    {
        // 리스트를 반으로 나누기
        List<int> leftWeights = weights.GetRange(0, N / 2);
        List<int> rightWeights = weights.GetRange(N / 2, N - N / 2);

        // 각각의 부분 집합의 합을 계산
        List<long> leftSubsets = GetSubsets(leftWeights);
        List<long> rightSubsets = GetSubsets(rightWeights);

        // 이진 탐색을 위한 정렬
        rightSubsets.Sort();

        long count = 0;

        // 왼쪽 부분 집합의 합을 순회하면서, C를 넘지 않는 조합을 찾기
        foreach (long leftSum in leftSubsets)
        {
            if (leftSum <= C)
            {
                // C - leftSum 보다 작거나 같은 값을 갖는 오른쪽 부분 집합의 개수를 구함
                int index = rightSubsets.BinarySearch(C - leftSum);
                
                if (index < 0)
                {
                    index = ~index; // BinarySearch가 실패하면 위치를 반환하기 위해 인덱스 변환
                }
                else
                {
                    // 정확히 C - leftSum인 값이 있는 경우, 동일한 값까지 포함시키기 위한 로직
                    while (index < rightSubsets.Count && rightSubsets[index] == C - leftSum)
                    {
                        index++;
                    }
                }

                count += index;
            }
        }

        return count;
    }

    static void Main()
    {
        // 한 줄에서 입력 받도록 수정
        var input = Console.ReadLine().Split(' ');
        int N = int.Parse(input[0]);
        long C = long.Parse(input[1]);

        List<int> weights = Console.ReadLine().Split(' ').Select(int.Parse).ToList();

        // 가능한 조합의 수를 계산하여 출력
        Console.WriteLine(CountValidCombinations(N, C, weights));
    }
}

File: BAEKJOON/1450 - 냅색문제/냅색문제.py
from itertools import combinations
from bisect import bisect_right

def get_subsets(weights):
    n = len(weights)  # 리스트의 길이(원소의 개수)를 변수 n에 저장
    subsets = []  # 부분 집합의 합을 저장할 빈 리스트 생성
    for i in range(n + 1):  # 부분 집합의 크기를 0부터 n까지 반복
        for comb in combinations(weights, i):  # 각 크기에 해당하는 모든 조합을 생성
            subsets.append(sum(comb))  # 조합의 합을 계산하여 subsets 리스트에 추가
    return subsets  # 계산된 부분 집합의 합 리스트를 반환

def count_valid_combinations(N, C, weights):
    left_weights = weights[:N//2]  # 리스트의 첫 번째 절반을 left_weights로 설정
    right_weights = weights[N//2:]  # 리스트의 나머지 절반을 right_weights로 설정

    left_subsets = get_subsets(left_weights)  # 왼쪽 절반의 부분 집합의 합 계산
    right_subsets = get_subsets(right_weights)  # 오른쪽 절반의 부분 집합의 합 계산

    right_subsets.sort()  # 이진 탐색을 위해 오른쪽 부분 집합의 합을 오름차순으로 정렬

    count = 0  # 가능한 조합의 수를 저장할 변수 초기화

    for left_sum in left_subsets:  # 왼쪽 부분 집합의 합을 하나씩 순회
        if left_sum <= C:  # 왼쪽 부분 집합의 합이 C 이하인 경우에만 처리
            count += bisect_right(right_subsets, C - left_sum)  
            # 오른쪽 부분 집합에서 left_sum과 합쳐서 C 이하가 되는 경우의 수를 이진 탐색으로 계산하여 count에 더함

    return count  # 최종 계산된 가능한 조합의 수 반환

N, C = map(int, input().split())  # 첫 번째 줄의 입력 값을 받아 N과 C에 저장
weights = list(map(int, input().split()))  # 두 번째 줄의 입력 값을 리스트로 받아 weights에 저장

print(count_valid_combinations(N, C, weights))  # 가능한 조합의 수를 계산하여 출력
